examples/ota-provider-app/esp32/main/main.cpp

```diff
@@ -0,0 +1,240 @@
+#include "esp_log.h"
+#include "esp_netif.h"
+#include "esp_spi_flash.h"
+#include "esp_system.h"
+#include "esp_wifi.h"
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "nvs_flash.h"
+#include <app-common/zap-generated/callback.h>
+#include <app/server/Server.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+#include <cmath>
+#include <cstdio>
+#include <string>
+#include <vector>
+
+#include <credentials/DeviceAttestationCredsProvider.h>
+#include <credentials/examples/DeviceAttestationCredsExample.h>
+
+#include <lib/support/ErrorStr.h>
+
+#include <app/clusters/ota-provider/ota-provider.h>
+#include <ota-provider-common/BdxOtaSender.h>
+#include <ota-provider-common/OTAProviderExample.h>
+
+using chip::BitFlags;
+using chip::app::Clusters::OTAProviderDelegate;
+using chip::bdx::TransferControlFlags;
+using chip::Callback::Callback;
+using chip::Messaging::ExchangeManager;
+using namespace ::chip;
+using namespace ::chip::System;
+using namespace ::chip::Credentials;
+using namespace ::chip::DeviceManager;
+using namespace ::chip::DeviceLayer;
+
+const char * TAG = "ota-provider-app";
+
+static DeviceCallbacks EchoCallbacks;
+
+// TODO: this should probably be done dynamically
+constexpr chip::EndpointId kOtaProviderEndpoint = 0;
+
+constexpr uint32_t kMaxBdxBlockSize                = 1024;
+constexpr chip::System::Clock::Timeout kBdxTimeout = chip::System::Clock::Seconds16(5 * 60); // Specifications mandates >= 5 minutes
+constexpr chip::System::Clock::Timeout kBdxPollFreq = chip::System::Clock::Milliseconds32(500);
+const char * gOtaFilename                           = "hello-world.bin";
+BdxOtaSender bdxServer;
+
+const esp_partition_t * otaPartition = nullptr;
+uint32_t otaImageLen                 = 0;
+uint32_t otaTransferInProgress       = false;
+
+void OnBlockQuery(void * context, chip::System::PacketBufferHandle & blockBuf, size_t & size, bool & isEof, uint32_t offset);
+void OnTransferComplete(void * context);
+void OnTransferFailed(void * context);
+
+chip::Callback::Callback<OnBdxBlockQuery> mOnBlockQuery(OnBlockQuery, nullptr);
+chip::Callback::Callback<OnBdxTransferComplete> mOnTransferComplete(OnTransferComplete, nullptr);
+chip::Callback::Callback<OnBdxTransferFailed> mOnTransferFailed(OnTransferFailed, nullptr);
+
+void OnBlockQuery(void * context, chip::System::PacketBufferHandle & blockBuf, size_t & size, bool & isEof, uint32_t offset)
+{
+    if (otaTransferInProgress == false)
+    {
+        if (otaPartition == nullptr || otaImageLen == 0)
+        {
+            ESP_LOGE(TAG, "OTA partition not found");
+            return;
+        }
+        otaTransferInProgress = true;
+    }
+
+    if (offset + kMaxBdxBlockSize >= otaImageLen)
+    {
+        size  = otaImageLen - offset;
+        isEof = true;
+    }
+    else
+    {
+        size  = kMaxBdxBlockSize;
+        isEof = false;
+    }
+
+    esp_err_t err = esp_partition_read(otaPartition, offset + sizeof(otaImageLen), blockBuf->Start(), size);
```

Comment:
[TransferSession:: GetTransferBlockSize()](https://github.com/project-chip/connectedhomeip/blob/e24e76ce1cbec8ee11fb8c4fc6f865faceae8d4f/src/protocols/bdx/BdxTransferSession.h#L279) will give you the negotiated block size value. However, the `TransferSession` object is private to `BdxDownloader` so you'll need to add an accessor.

Reply:
Checking the length of the PacketBuffer's data won't do the right thing either?  As in, there can be trailing data of some sort?

Reply:
What is the purpose of `offset + sizeof(otaImageLen)`?

Reply:
Since the OTA image resides in the flash, we do not know the it's size. The specification defined the header for this purpose but, I thought for now using just the size of image would be fine. So, `otaImageLen` is a four bytes at the start of OTA image. And purpose of `offset + sizeof(otaImageLen)` is just to skip the first four bytes.